### PR TITLE
Fix PATH and git-related build issues on Jenkins

### PR DESCRIPTION
**Pull request by dividedmind**
There were a couple problems preventing builds from working. 

One was that Jenkins git plugin recently changed to default to `git fetch --no-tags`, which broke version tagging. Quite a lot of digging in Java code later, I discovered that Jenkins git plugin is fundamentally broken WRT `git fetch`: it only allows using `--tags` or `--no-tags`; of the two the latter is definitely a better default, but even better would be to use neither of these options and use default git behaviour of fetching tags that are included in what is being fetched already. Fortunately it's a simple matter of running `git fetch` to fetch the missing tags.

Another issue was that PATH went mysteriously missing from the init process. I'm not entirely sure what is the exact cause of that change, but it was a good excuse to make the whole build and boot process a little more robust. Now environment and command line are imported from docker image metadata on conversion, which ensures that the runtime will work correctly even if we change things in the appliance. (It's still not perfect in that regard, but it's a lot better.)

Initially I wanted to make some changes in evoke to explicitly set PATH in runit scripts -- something which might be a good idea anyway -- but this would have only masked the issue with this build.

Long story short, Jenkins build is green: https://jenkins.conjur.net/job/conjurinc--appliance-uml/job/bugfix%2Fjenkins-build-issues/21/

Please tag v0.4.1 after merging, or let me know and I will.

### CODE REVIEW

File: imgcvt/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,7 @@
 FROM alpine
 
 RUN apk update && apk add squashfs-tools libarchive-tools && rm -rf /var/cache/apk/*
+RUN mkdir /tmp/root
```
Comment (dustinmm80) [inline imgcvt/Dockerfile]: Can we call this `/tmp/docker` or something to match more closely what this dir is?

File: mkimage.sh [hunk 2]
```diff
@@ -3,21 +3,21 @@
 SRC_IMAGE="$1"
 [ -z "$SRC_IMAGE" ] && ( echo "image argument required"; exit 1 )
 
-CVT_IMAGE=imgcvt:`cat imgcvt/* | shasum -a 256 | cut -c-16`
 
-docker pull $SRC_IMAGE
+has_image()
+{
+  docker inspect --type image $1 > /dev/null 2>&1
```
Comment (dustinmm80) [inline mkimage.sh]: if we don't need the output of `inspect` we shouldn't `/dev/null 2>&1` IMO. That throws all output away, makes build issues hard to understand.

`docker images | grep "$1"` would work too and preserve output.

File: mkimage.sh [hunk 3]
```diff
@@ -32,7 +32,18 @@ else
 fi
 
 echo Converting in $CONTAINER-cvt...
-docker export $CONTAINER | $PV | docker run --name $CONTAINER-cvt -i $CVT_IMAGE
+docker create --name $CONTAINER-cvt -i $CVT_IMAGE
```
Comment (dustinmm80) [inline mkimage.sh]: Feels like we should use docker-compose for this, but we can get that on a future enhancement.

File: scripts/launch.sh [hunk 4]
```diff
@@ -68,6 +68,6 @@ echo ""
 
 export SLIRP IP {SSH,HTTPS,LDAPS,POSTGRES,PG_ARCHIVE}_PORT TMPDIR
 
-cmdline="$KERNEL con0=null,fd:1 ubd0=$IMAGE ubd1=$DATAVOL initrd=$INITRAMFS eth0=slirp,,$SETSLIRP internal_ip=$IP pubkey=$PUBKEY mem=$MEM $args"
-echo Running with: $cmdline
-exec $cmdline
+cmdline=($KERNEL con0=null,fd:1 ubd0=$IMAGE ubd1=$DATAVOL initrd=$INITRAMFS eth0=slirp,,$SETSLIRP internal_ip=$IP pubkey=$PUBKEY mem=$MEM "${args[@]}")
```
Comment (dustinmm80) [inline scripts/launch.sh]: This is all a bit confusing, can you try to make it more clear what is happening here? I usually try to avoid `exec` in bash, it's hard to follow sometimes.

File: imgcvt/Dockerfile (hunk 1 above)
Comment (dividedmind) [reply imgcvt/Dockerfile]: It matches it more closely than `/tmp/docker` IMO: it's the root of the filesystem image. I'm not sure what `/tmp/docker` would mean. I also don't believe it's particularly important that this path is descriptive, it's pretty much local to `cvt.sh`. I suppose I could make it configurable via envar in Dockerfile, but YAGNI.

File: mkimage.sh (hunk 2 above)
Comment (dividedmind) [reply mkimage.sh]: In my opinion `docker images | grep` would be a very bad idea, even if it worked. Here is why:
1. `docker images` is human-readable, not machine readable. Parsing human-readable output leads to insanity.
2. `docker images` doesn't list all images in `image:tag` format. I suppose I could write a parser for docker image tags in awk, then transform that to a grep re, which parses the human-readable table. I would also have to consider when tag is not given, then substitute `latest`. Oh, and what if someone specifies the image by sha, like `image@sha256:deadbeef`? Or image id? Or whatever docker comes up with next week? I would need to support these as well.
3. Still it's human-readable output. I could set `LC_ALL=C` to at least try to prevent the aforementioned script (which would probably be a separate utility, maybe I should write it in another language? `bash` is cumbersome for non-trivial text processing) from getting tripped on language-related format differences. Let's keep our fingers crossed that nobody adds a new column or something.
4. Let's just make sure we have the right whitespace anchors in our RE. We don't want `cafebabe` to match part of an image hash (well, unless that's what the user intended, maybe it's a hash prefix?), or `test` to match `nuclear-test`. Or a `test` tag in `death-star` image.
5. I suppose at this point it would be easier to just call docker HTTP API directly with curl. This would do away with the problems with 3 and 4 above. But then let's not hardcode the endpoint URL, lest it only works with local docker instance.

If only there was a tool that's already written, and has these properties:
- is easily available,
- parses image specifiers with exact Docker conventions,
- uses system Docker configuration,
- just checks whether the image is available or not (whether to it being missing or a problem with Docker daemon) and tells us the answer.

Oh, wait. That's _exactly_ what `docker inspect --type $id` does, if we throw away all its output.

/rant ;)

File: mkimage.sh (hunk 3 above)
Comment (dividedmind) [reply mkimage.sh]: I don't really like using docker-compose for automation. It's too much magic, which is great for dev workflows, but leads to very brittle CI code. Also `docker-compose` doesn't really fit well here IMO even if it wasn't for that; you can certainly make it work if you were so inclined, but why use the wrong tool for the job?

File: scripts/launch.sh (hunk 4 above)
Comment (dividedmind) [reply scripts/launch.sh]: I'm building a command line, and I'm gathering it in a variable like that only so I can print it. Good point, I suppose it'd be clearer to instead just `set -x` before running it; I'll do that, thanks!

Can you tell me more about why do you feel `exec` is hard to follow? It's strictly simpler than the same line without `exec` (_execve(2)_ vs _fork(2)_-then-_execve(2)_).

**Comment by dustinmm80** [comment]
Looks good now, thanks Rafal!

Once https://jenkins.conjur.net/job/conjurinc--appliance-uml/job/bugfix%252Fjenkins-build-issues/25/ is green I'll approve and merge.

File: scripts/launch.sh [hunk 5]
```diff
@@ -68,6 +68,14 @@ echo ""
 
 export SLIRP IP {SSH,HTTPS,LDAPS,POSTGRES,PG_ARCHIVE}_PORT TMPDIR
 
-cmdline="$KERNEL con0=null,fd:1 ubd0=$IMAGE ubd1=$DATAVOL initrd=$INITRAMFS eth0=slirp,,$SETSLIRP internal_ip=$IP pubkey=$PUBKEY mem=$MEM $args"
-echo Running with: $cmdline
-exec $cmdline
+set -x
+exec $KERNEL \
```
Comment (dustinmm80) [inline scripts/launch.sh]: thanks, much better!

File: mkimage.sh (hunk 3 above)
Comment (dustinmm80) [reply mkimage.sh]: Good points, let's leave it how it is now that it's green again.

File: mkimage.sh (hunk 2 above)
Comment (dustinmm80) [reply mkimage.sh]: TIL more about `docker inspect --type $id`, thanks Rafal!

I'll use this more when it seems to fit in other code bases.

**Comment by dividedmind** [comment]
Thanks! And sorry for the rant, I've gotten carried away a bit. As my excuse I'll say that I probably did put a lot more thought than it was worth it into how to best implement `has_image()`, and IIRC `images | grep` was my first idea, so I had this whole argument precomputed just waiting to come out ;)

File: mkimage.sh (hunk 2 above)
Comment (dividedmind) [reply mkimage.sh]: (Just to clarify: `--type image $id`; my typo. (Without `--type image` it would also try to find containers matching `$id`.))
